elements/lisk-transactions/src/13_vote_transaction.ts

```diff
@@ -0,0 +1,439 @@
+	public readonly asset: VoteAsset;
+	public static TYPE = 13;
+
+	public constructor(rawTransaction: unknown) {
+		super(rawTransaction);
+		const tx = (typeof rawTransaction === 'object' && rawTransaction !== null
+			? rawTransaction
+			: {}) as Partial<TransactionJSON>;
+		if (tx.asset) {
+			const rawAsset = tx.asset as RawAsset;
+			this.asset = {
+				votes: rawAsset.votes.map(vote => {
+					const amount = isNumberString(vote.amount)
+						? BigInt(vote.amount)
+						: BigInt(0);
+
+					return {
+						delegateAddress: vote.delegateAddress,
+						amount,
+					};
+				}),
+			};
+		} else {
+			// tslint:disable-next-line no-object-literal-type-assertion
+			this.asset = { votes: [] };
+		}
+	}
+
+	public assetToJSON(): object {
+		return {
+			votes: this.asset.votes.map(vote => ({
+				delegateAddress: vote.delegateAddress,
+				amount: vote.amount.toString(),
+			})),
+		};
+	}
+
+	protected assetToBytes(): Buffer {
+		const bufferArray = [];
+		for (const vote of this.asset.votes) {
+			const addressBuffer = intToBuffer(
+				vote.delegateAddress.slice(0, -1),
+				SIZE_UINT64,
+			);
+			bufferArray.push(addressBuffer);
+			const amountBuffer = intToBuffer(
+				vote.amount.toString(),
+				SIZE_INT64,
+				'big',
+				true,
+			);
+			bufferArray.push(amountBuffer);
+		}
+
+		return Buffer.concat(bufferArray);
+	}
+
+	public async prepare(store: StateStorePrepare): Promise<void> {
+		const addressArray = this.asset.votes.map(vote => ({
+			address: vote.delegateAddress,
+		}));
+		const filterArray = [
+			{
+				address: this.senderId,
+			},
+			...addressArray,
+		];
+
+		await store.account.cache(filterArray);
+	}
+
+	protected validateAsset(): ReadonlyArray<TransactionError> {
+		const asset = this.assetToJSON();
+		const schemaErrors = validator.validate(voteAssetFormatSchema, asset);
+		const errors = convertToAssetError(
+			this.id,
+			schemaErrors,
+		) as TransactionError[];
+
+		// tslint:disable-next-line no-let
+		let upvoteCount = 0;
+		// tslint:disable-next-line no-let
+		let downvoteCount = 0;
+		const addressSet = new Set();
+		for (const vote of this.asset.votes) {
+			addressSet.add(vote.delegateAddress);
+			if (vote.amount > BigInt(0)) {
+				upvoteCount += 1;
+			} else if (vote.amount < BigInt(0)) {
+				downvoteCount += 1;
+			} else {
+				errors.push(
+					new TransactionError(
+						'Amount cannot be 0',
+						this.id,
+						'.asset.votes.amount',
+					),
+				);
+			}
+			if (vote.amount % TEN_UNIT !== BigInt(0)) {
+				errors.push(
+					new TransactionError(
+						'Amount should be multiple of 10 * 10^8',
+						this.id,
+						'.asset.votes.amount',
+						vote.amount.toString(),
+					),
+				);
+			}
+		}
+		if (upvoteCount > MAX_VOTE) {
+			errors.push(
+				new TransactionError(
+					'Upvote can only be casted upto 10',
+					this.id,
+					'.asset.votes',
+					upvoteCount,
+					'10',
+				),
+			);
+		}
+		if (downvoteCount > MAX_VOTE) {
```

Comment:
don't we need the check for `upvoteCount + downvoteCount < 21`  i.e. no more than 20 votes in total?

Reply:
total entry check is done by schema =)

Reply:
Ahh the schema will catch it I think. Resolving.